scripts/generators/typescript.js

```diff
@@ -0,0 +1,155 @@
+"use strict";
+
+const t = require("../../packages/babel-types");
+const utils = require("./utils");
+
+const NODE_PREFIX = "";
+
+let code = `// NOTE: This file is autogenerated. Do not modify.
+// See scripts/generators/typescript.js for script used.
+
+export class ${NODE_PREFIX}Comment {
+  value: string;
+  start: number;
+  end: number;
+  loc: ${NODE_PREFIX}SourceLocation;
+}
+
+export class ${NODE_PREFIX}BlockComment extends ${NODE_PREFIX}Comment {
+  type: "BlockComment";
+}
+
+export class ${NODE_PREFIX}LineComment extends ${NODE_PREFIX}Comment {
+  type: "LineComment";
+}
+
+export class ${NODE_PREFIX}SourceLocation {
+  start: {
+    line: number;
+    column: number;
+  };
+
+  end: {
+    line: number;
+    column: number;
+  };
+}
+
+export class Node {
+  leadingComments: Array<${NODE_PREFIX}Comment> | null;
+  innerComments: Array<${NODE_PREFIX}Comment> | null;
+  trailingComments: Array<${NODE_PREFIX}Comment> | null;
+  start: number | null;
+  end: number | null;
+  loc: ${NODE_PREFIX}SourceLocation | null;
+}\n\n`;
+
+//
+
+const lines = [];
+
+for (const type in t.NODE_FIELDS) {
+  const fields = t.NODE_FIELDS[type];
+
+  const struct = ['type: "' + type + '";'];
+  const args = [];
+
+  Object.keys(t.NODE_FIELDS[type])
+    .sort((fieldA, fieldB) => {
+      const indexA = t.BUILDER_KEYS[type].indexOf(fieldA);
+      const indexB = t.BUILDER_KEYS[type].indexOf(fieldB);
+      if (indexA === indexB) return fieldA < fieldB ? -1 : 1;
+      if (indexA === -1) return 1;
+      if (indexB === -1) return -1;
+      return indexA - indexB;
+    })
+    .forEach(fieldName => {
+      const field = fields[fieldName];
+      const typeAnnotation = utils.stringifyValidator(
+        field.validate,
+        NODE_PREFIX
+      );
+
+      args.push(
+        t.toBindingIdentifierName(fieldName) +
+          (isNullable(field) ? "?: " : ": ") +
+          typeAnnotation
+      );
+
+      if (t.isValidIdentifier(fieldName)) {
+        struct.push(
+          fieldName +
+            ": " +
+            typeAnnotation +
+            (isNullable(field) ? " | null" : "") +
+            ";"
+        );
+      }
+    });
+
+  code += `export class ${NODE_PREFIX}${type} extends Node {
+  ${struct.join("\n  ").trim()}
+}\n\n`;
+
+  // super and import are reserved words in JavaScript
+  if (type !== "Super" && type !== "Import") {
+    lines.push(
+      `export function ${utils.toFunctionName(type)}(${args.join(
+        ", "
+      )}): ${NODE_PREFIX}${type};`
+    );
+  }
+}
+
+for (let i = 0; i < t.TYPES.length; i++) {
+  let decl = `export function is${
+    t.TYPES[i]
+  }(node: object, opts?: object | null): `;
+
+  if (t.NODE_FIELDS[t.TYPES[i]]) {
+    decl += `node is ${NODE_PREFIX}${t.TYPES[i]};`;
+  } else {
+    decl += `boolean;`;
+  }
+
+  lines.push(decl);
+}
+
+lines.push(
+  `export function validate(n: BabelNode, key: string, value: mixed): void;`,
+  `export function clone<T>(n: T): T;`,
```

Comment:
This is missing the second argument.

`T` needs to `extend Node`, and the second argument should be `Partial<T>` (lazy typing), or a hack similar to what is used for `setState` in `@types/react` should be used to correctly forbid unintentionally using `undefined` as a value.